### PR TITLE
Add a Request.wait_for_disconnection() method

### PR DESCRIPTION
**Pull request by gjcarneiro**
as means of allowing request handlers to be notified of premature client
disconnections.

## What do these changes do?

Adds a new async method `Request.wait_for_disconnection()`.  This method does nothing except to wait until the connection that sent a request to be disconnected.

This may be useful, in certain cases, now that aiohttp no longer automatically cancels request handlers upon client disconnection.  E.g., if there is a slow DB query, it is beneficial to be able to cancel that DB query as soon as the client disconnects, to avoid the DB doing useless work.

This async method provides the needed "signal" for that.  Although making use of it is not pretty, it can be simplified by a utility function/wrapper, probably outside the scope of aiohttp (and outside the scope of this PR).

## Are there changes in behavior for the user?

A new method was added, as described above.  It is opt-in, so should not affect existing code.

## Related issue number

#2492 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by gjcarneiro** [comment]
One check is failing because of:

> The job running on agent Azure Pipelines 7 has exceeded the maximum execution time of 10. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134
> The operation was canceled.

I don't think it's my fault that it's failing.  Possibly the CI environment is broken.  Limiting to 10 minutes seems rather restrictive...

**Comment by asvetlov** [comment]
Yes, tests hang on MacOS periodically.
I'm working on the fix. It is reproduced on a relative heavy load.

File: aiohttp/web_request.py [hunk 1]
```diff
@@ -690,6 +694,21 @@ def __bool__(self) -> bool:
 
     def _cancel(self, exc: BaseException) -> None:
         self._payload.set_exception(exc)
+        for fut in self._disconnection_waiters:
+            fut.set_result(None)
```
Comment (asvetlov) [inline aiohttp/web_request.py]: Welcome to low-level asyncio!
This line contains a very common but annoying bug. Let me describe the failure scenario:
1. a task calls `await request.wait_for_disconnection()` and waits
2. the task gets canceled by explicit `task.cancel()` call
3. eventually the connection is broken by peer. `request._cancel()` is called; which in turn calls `fut.set_result(None)`.
4. The future is canceled by bullet 2, `set_result()` fails.

Please use `helpers.set_result()`.

**Review by asvetlov — CHANGES_REQUESTED** [review]
Please fix my note and update the documentation.

`.. versionadded::` should use `4.0` as you use 4.0-specific `request._cancel()`

File: aiohttp/web_request.py (hunk 1 above)
Comment (gjcarneiro) [reply aiohttp/web_request.py]: Ah, good catch.  Changed to use helpers.set_result.

Comment (gjcarneiro) [reply aiohttp/web_request.py]: versionadded...
Note: this PR needs rebase/squash commits before merging.

**Review by asvetlov — CHANGES_REQUESTED** [review]
1. A client opens a connection in HTTP keep-alive mode
2. server code calls `await request.wait_for_disconnection()`
3. HTTP request is handled *without* socket closing (keep-alive mode is on)
4. `await request.wait_for_disconnection()` hangs forever

Please fix it

**Comment by gjcarneiro** [comment]
> 1. A client opens a connection in HTTP keep-alive mode
> 2. server code calls `await request.wait_for_disconnection()`
> 3. HTTP request is handled _without_ socket closing (keep-alive mode is on)
> 4. `await request.wait_for_disconnection()` hangs forever
> 
> Please fix it

The method name clearly says "wait for disconnection".  In case "HTTP request is handled _without_ socket closing (keep-alive mode is on)" there is no disconnection. That is a use case I expected.

The way I envision this to be used, then `request.wait_for_disconnection()` coroutine call would always be cancelled at the end of each request, by the app code.  This is meant to be just a building block for something more convenient to use, like a decorator.  From the code I posted on the issue, I can create a decorator like:

```python

def cancel_on_disconnect(handler):
  @functools.wraps(handler)
  async def wrap_handler(request):
     wait = asyncio.create_task(request.wait_for_disconnection())
     bgtask = asyncio.create_task(handler())
     asyncio.wait([wait, bgtask], return_when=asyncio.FIRST_COMPLETED)
     if bgtask.done():
         wait.cancel()
         return web.Response(bgtask.result())
     bgtask.cancel()
     raise web.HTTPGatewayTimeout()
  return wrap_handler

@cancel_on_disconnect
def my_handler(request):
   ...
```

The above code probably has bugs, I didn't even test it.  But the point is that `request.wait_for_disconnection()` will _always_ be cancelled at the end of the request anyway.

I can try to make it return at request end, even if there is no disconnection due to keepalive, but honestly I don't think it matters.

**Comment by asvetlov** [comment]
I have a feeling that time of `request.wait_for_disconnection()` waiting should be less than the lifetime of request object itself.
Cancelling all pending waiters after web-handler finishing makes API cleaner and safer to use, isn't it?

**Comment by gjcarneiro** [comment]
OK, _fine_...  but then the documentation, and maybe method name, might need to be updated to reflect that the function also returns at the end of request, not just client disconnection.

**Comment by asvetlov** [comment]
I was talking about cancellation, not a regular `None` return.
Documentation should reflect it, sure.
Regarding the method name: do you have a proposal?

**Comment by gjcarneiro** [comment]
> I was talking about cancellation, not a regular None return.

I didn't understand this, but makes sense.

> Regarding the method name: do you have a proposal?

Because it cancels it, there is no need for renaming, I think.

**Comment by asvetlov** [comment]
looks good. The last remaining part is documentation update

File: aiohttp/web_request.py [hunk 2]
```diff
@@ -690,6 +695,27 @@ def __bool__(self) -> bool:
...
+        for fut in self._disconnection_waiters:
+            fut.cancel()
+
+    async def wait_for_disconnection(self) -> None:
+        """Returns when the connection that sent this request closes
+
+        This can be used in handlers as a means to receive a notification of
+        premature client disconnection.
+
+        .. versionadded:: 4.0
```
Comment (asvetlov) [inline aiohttp/web_request.py]: It should be in `./docs/web_reference.rst`, not in python sources.

**Comment by gjcarneiro** [comment]
Note: feel free to rebase/squash commits, or else I can do it.

File: aiohttp/web_request.py [hunk 3]
```diff
@@ -690,6 +695,20 @@ def __bool__(self) -> bool:
...
         self._payload.set_exception(exc)
+        for fut in self._disconnection_waiters:
+            set_result(fut, None)
+
+    def _finish(self) -> None:
+        for fut in self._disconnection_waiters:
+            fut.cancel()
+
+    async def wait_for_disconnection(self) -> None:
+        fut = asyncio.Future()  # type: asyncio.Future[None]
```
Comment (asvetlov) [inline aiohttp/web_request.py]:
```suggestion
        loop = asyncio.get_event_loop()
        fut = loop.create_future()  # type: asyncio.Future[None]
```

**Review by asvetlov — CHANGES_REQUESTED** [review]
Please apply the suggestion.

No rebase is required; we use squash-on-merging strategy for aiohttp RPs

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
Thanks!

**Pull request merged**